Fix state "svn.latest" diff output in test mode

sebw opened the pull request:
### What does this PR do?

It fixes incorrect returned info from `svn.latest` when run in test mode.

### What issues does this PR fix or reference?

It fixes https://github.com/saltstack/salt/issues/41022

### Previous Behavior

Problematic scenario:

- current checked out code on minion is at revision 1
- we define the desired revision `rev: 2` in our svn.latest state
- latest revision on the SVN repository has already reached revision 3

In test mode, the output would show the diff between revision 3 and "working copy" (1), instead of showing a diff between revision 1 and 2:

```
----------
          ID: svn_repo
    Function: svn.latest
        Name: file:///home/svn
      Result: None
     Comment: Index: /var/www/svn/README.md
              ===================================================================
              --- /var/www/svn/README.md	(revision 3)
              +++ /var/www/svn/README.md	(working copy)
              @@ -1,3 +1,3 @@
               initial import
               -------------
              -rev 3
              +rev 1
     Started: 22:11:01.241382
    Duration: 18.837 ms
     Changes:
```

This output presents two other problems:

- diff is confusing giving the idea that text "rev 1" will be added to README.md while it is already there
- output is not giving an idea of the current checked out revision, just "working copy"

The proposed output:

```
----------
          ID: svn_repo
    Function: svn.latest
        Name: file:///home/svn
      Result: None
     Comment: Index: /var/www/svn/README.md
              ===================================================================
              --- /var/www/svn/README.md	(revision 1)
              +++ /var/www/svn/README.md	(revision 2)
              @@ -1,3 +1,3 @@
               initial import
               -------------
              -rev 1
              +rev 2
     Started: 22:11:01.241382
    Duration: 18.837 ms
     Changes:
```

This output is better:

- it compares against the desired revision
- it shows current version number instead of "working copy"
- diff is actually showing what will happen when state is run

What was the problem:

`-r HEAD` was hardcoded in the test block section instead of comparing with `-r current_rev:new_rev`.

### New Behavior

svn.latest now compares current code against the desired revision.

### Tests written?

No

### Additional note

This is my first PR, I've read the guidelines, but please bear with me if I made an obvious mistake :)

whiteinge committed:
We always want the token_tool enabled; it only performs bookkeeping

whiteinge committed:
Update reference JavaScript app

Don't use cookies for auth. Better app structure that doesn't rely on
eventsource hacks to show login page and other improvements.

sebw committed:
SVN diff checking against revision instead of HEAD

When revision was set in state, svn.latest with test=True was checking against HEAD instead of defined revision.

We are now checking current revision against specified revision in state.

Fixes https://github.com/saltstack/salt/issues/41022

sebw committed:
SVN diff checking against revision instead of HEAD

When revision was set in state, svn.latest with test=True was checking against HEAD instead of defined revision.

We are now checking current revision against specified revision in state.

Fixes https://github.com/saltstack/salt/issues/41022

sebw committed:
Removing useless str()

sebw committed:
Removing useless str()

Commit:
Add wildcard support initial class

Commit:
Add unit test scaffold for wildcard support

gtmanfred committed:
add resolving extra flags to yum upgrade

salt-jenkins commented:
@sebw, thanks for your PR! By analyzing the history of the files in this pull request, we identified @UtahDave, @Talkless and @ruebenramirez to be potential reviewers.

sebw committed:
Fixed indentation

sebw committed:
Fixed indentation

Commit:
Fix: zypper handling of multiple version packages

For kernel packages, zypper reports multiple versions as "new" during
upgrade. Thus, a salt state requesting pkg.latest of any kernel package
would fail, even though zypper runs successfully. This commit provides
a simple fix.

Commit:
Rename UT to a proper name

Commit:
Fix a docstring

Commit:
Add UT for matching all versions

Commit:
Add UT to match multiple asterisks

Commit:
Add UT for exact match of the pattern at the end

Commit:
Add UT for exact match of the pattern at the beginning

Commit:
Test the exact methods for matcher

Commit:
Add UT for wildcard finder usage

Commit:
Add import for LooseVersion dependency

Commit:
Implement wildcard converter

gtmanfred committed:
use six and clean_kwargs

cachedout commented on salt/states/svn.py:
```diff
@@ -95,13 +95,27 @@ def latest(name,
+    try:
+        current_info = __salt__['svn.info'](cwd, target, user=user, username=username, password=password, fmt='dict')
+        svn_cmd = 'svn.update'
+    except exceptions.CommandExecutionError:
```
If we catch this, won't `svn_cmd` not be updated and we'll do the wrong thing later on?

sebw replied:
Hey Mike,

It was not problematic in itself (svn_cmd was declared before the exception and it would do the right thing if exception was caught), but while reviewing my code, I discovered another bug when the target directory existed but was not a svn working copy.

I added an exception statement within the test block now, it should make the code clearer.

If target directory exists and is not a working copy, it fails:

```
          ID: svn_repo
    Function: svn.latest
        Name: file:///home/svn
      Result: False
     Comment: /var/www/svn exists but is not a svn working copy.
     Started: 10:43:08.289164
    Duration: 17.355 ms
     Changes:
```

If the directory doesn't exist and is set to be checked out, it now returns the revision it is going to check out:

```
          ID: svn_repo
    Function: svn.latest
        Name: file:///home/svn
      Result: None
     Comment: /var/www/svn doesn't exist and is set to be checked out at revision 4.
     Started: 10:35:23.961889
    Duration: 0.498 ms
     Changes:
```

If directory is set to be updated it gives:

```
          ID: svn_repo
    Function: svn.latest
        Name: file:///home/svn
      Result: None
     Comment: Index: /var/www/svn/README.md
              ===================================================================
              --- /var/www/svn/README.md	(revision 3)
              +++ /var/www/svn/README.md	(revision 4)
              @@ -1,3 +1,3 @@
               initial import
               -------------
              -rev 3
              +rev 4
     Started: 10:45:42.755722
    Duration: 34.862 ms
     Changes:
```

Commit:
Add basic auth for SPM

sebw committed:
Better checks around target directories

sebw committed:
Better checks around target directories

BenoitKnecht committed:
states: cron: show correct changes when using `special`

When using the `cron.present` state with the `special` keyword:

  /path/to/cron/script:
    cron.present:
      - user: root
      - special: '@hourly'

and running Salt in test mode, Salt would always report changes, even
when nothing needed to be changed.

This commit actually checks if changes are needed when using the
`special` keyword.

abednarik committed:
Update apt module regarding upgrade against hold packages.

Fixes #41231 and related to #30777.

Commit:
Merge pull request #41181 from gtmanfred/2016.11

add resolving extra flags to yum upgrade

Commit:
Bugfix: unable to use 127 as hostname

Commit:
Unit test for accepting hosts names as 127

Commit:
Harden to 127. IP part

Commit:
Add unit test for hostname can be started from 127

Commit:
Merge pull request #41283 from terminalmage/bp-41251

Backport #41251 to 2016.11

Commit:
Added unit tests and copied the behavior to
.upgrade method, too.

Commit:
Compacted with statements in the unit test.

garethgreenaway committed:
Updating the code that is pulling in the list of cached minions to use self.cache.list instead of relying on checking the local file system, which only works for the localfs cache method.  #40748

whiteinge committed:
Make tools a little easier to read

garethgreenaway committed:
Removing chdir as it is no needed with this change

garethgreenaway committed:
Swapping the order in the func_alias so the ls function is available.

whiteinge committed:
Absorb CORS handler into CORS tool; allows OPTIONS to be unauthenticated

Tools can't short-circuit other tools so we need a little hack to avoid
the auth_tool check.

whiteinge committed:
No need to send a session cookie when GET'ing the root URL

Best not to send this to JavaScript SPAs when loading for the first time
since XmlHttpRequest will use it by default (fetch is an opt-in though).

whiteinge committed:
Add enable_sessions setting to disable all session-respecing URLs

whiteinge committed:
Add /token convenience endpoint

Commit:
Addition of powershell_all function

Commit:
Lint fix

Commit:
Reverted back to cascading with statements for python 2.6 compat

lomeroe committed:
properly convert packed string to decimal values

Commit:
Merge pull request #41303 from lomeroe/bp-41301

backport #41301 -- properly convert packed string to decimal values

Commit:
Merge pull request #41289 from garethgreenaway/2016_11_5_fix_consul_cache_ls

Fixing consul cache

Commit:
Merge pull request #41269 from isbm/isbm-minion-id-127-name

Bugfix: Unable to use "127" as hostname for the Minion ID

Commit:
Merge pull request #41236 from BenoitKnecht/2016.11

states: cron: show correct changes when using `special`

Commit:
change outdict to result

Commit:
Merge pull request #41232 from axmetishe/2016.11

Add basic auth for SPM

garethgreenaway committed:
Allowing test=True to be passed for salt.runner and salt.wheel when used with orchestration

Commit:
More changes
Added encoded_cmd to run_all.
Changed behaviour of powershell_all. No longer considers
Non-0 retcodes to be failures because some Windows executables
exit with non-0 on success. If a python dictionary can't be created,
the result is set to "False" instead of "None".
Expanded the description of retcode in the documentation.

Commit:
lint fix

Commit:
Fix typo

Commit:
Lint fix

Commit:
Lint fix

Ch3LL committed:
[2016.11] Bump latest release version to 2016.11.5

lomeroe committed:
backport 41307 to 2016.11, properly pack version numbers into single
number

Commit:
Merge pull request #41317 from Ch3LL/update_latest_2016.11

[2016.11] Bump latest release version to 2016.11.5

Commit:
Merge pull request #41208 from pkazmierczak/pkazmierczak-zypper-multiple-ver-pkgs

Fix: zypper handling of multiple version packages

Ch3LL committed:
Add patched packages warning to 2016.11.5 release notes

Ch3LL committed:
fix url refs in rst

Ch3LL committed:
Add 2016.11.6 Release Notes

lorengordon committed:
Corrects versionadded for win_network.get_route

tkwilliams committed:
INFRA-4842 - boto_elb should return None when test=True and changes are pending...

Commit:
Merge pull request #41322 from Ch3LL/fix_release_2016.11.5_notes

Add patched packages warning to 2016.11.5 release notes

tkwilliams committed:
INFRA-4842 - update unit test to account for new test=True result

aneeshusa committed:
Add TLS connection options for pgjsonb returner

Previously, any TLS options in the configuration were silently ignored,
and would not have worked regardless
because they were copied from the MySQL returner.

tkwilliams committed:
Revert "INFRA-4842 - boto_elb should return None when test=True and changes are pending..."

This reverts commit ae074faa12b3f85b370fcff086f79529e1a69d8c.

Commit:
Merge pull request #41329 from lorengordon/doc-fix

Corrects versionadded for win_network.get_route

Commit:
Merge pull request #41327 from Ch3LL/add_2016.11.6_release

Add 2016.11.6 Release Notes

Commit:
Merge pull request #41319 from lomeroe/bp_41307

backport #41307 to 2016.11, properly pack version numbers into single

Commit:
Merge pull request #41309 from garethgreenaway/38894_allowing_test_argument

Adding test argument for runners & wheel orchestration modules

Commit:
Merge pull request #41287 from garethgreenaway/40748_2016_11_consul

Fix to consul cache

Commit:
Made the changes I proposed

Commit:
empty stdout now results in CmdExecutionError exception

Commit:
Documentation fix

Commit:
Allow the polymorophism again, but with force_list paramater

sebw commented:
@cachedout if approved, could this bug be back-ported to the next point release?

whiteinge committed:
Lint fix

Not really a fan of this rule...

whiteinge committed:
Add full_return to release notes

whiteinge committed:
Add rest_cherrypy additions to the release notes

Commit:
Merge branch '2016.11' into 'nitrogen'

Conflicts:
  - salt/modules/yumpkg.py
  - salt/spm/__init__.py
  - tests/unit/modules/test_zypper.py
  - tests/unit/utils/test_network.py

Commit:
Returns None on no output now. "result_non_reason" field added to return
dict.

Commit:
Do it Damon Atkins way (mostly)

Commit:
Pylint fix

art-schneider committed:
Merge remote-tracking branch 'origin/develop' into develop

Commit:
Merge pull request #41358 from rallytime/merge-nitrogen

[nitrogen] Merge forward from 2016.11 to nitrogen

Commit:
Merge pull request #41356 from whiteinge/doc-full_return

Add full_return to release notes

Commit:
Merge pull request #41352 from whiteinge/rest_cherrypy-cors-token

rest_cherrypy CORS fix & /token endpoint

cachedout commented:
@sebw Yes, certainly.

Before we merge this, though, could you rebase it to the head of develop? I would like to get a clean test run here.

Commit:
Lint fixes

Commit:
Merge remote-tracking branch 'upstream/develop' into salt-states-svn-patch-41022

tkwilliams committed:
INFRA-4842 - fix issues with previous test=True return values

sebw commented:
@cachedout I'm not really comfortable with rebasing yet. I tried merging upstream/develop into my branch but Jenkins tests still fail. Can you point me to any doc explaining rebasing as you expect me to do it? Thanks!

tkwilliams committed:
INFRA-4842 - '==' != '=', dammit!

tkwilliams committed:
INFRA-4842 - fixup unit tests to suit

grahamhayes committed:
Add unit test for PR #40905

This ensures we do not introduce a regression which breaks
the usage of the `batch` option when using states in orchestration
runners.

Related to #40635 and #39169

Commit:
Merge branch 'nitrogen' into 'develop'

Conflicts:
  - salt/modules/zypper.py
  - tests/unit/utils/test_network.py

tkwilliams committed:
INFRA-4842 - /thought/ I fixed this :)

Commit:
Add "KeyError" to bare except in azurearm

Fixes pylint, refs #41346.

Commit:
Merge pull request #41395 from rallytime/lint-develop

Add "KeyError" to bare except in azurearm

Commit:
Don't try to split a dict object in zypper module

art-schneider committed:
Merge branch 'develop' of https://github.com/saltstack/salt into develop

art-schneider committed:
Don't create cron for cron.rm_job if no cron present for user

Commit:
nilrt_ip: fix examples

Fix examples for ip.set_dhcp_linklocal_all and ip.set_static_all

Signed-off-by: Heghedus Razvan <razvan.heghedus@ni.com>

Commit:
Return string type and only when version is specified

Commit:
Add UT for None type verification if version has been not passed

Commit:
Add UT for typecheck on the output data

Commit:
Add wildcard support to Zypper

Commit:
Add operator support for version wildcard

Commit:
Add UT for operator conditions

Commit:
Update docstring to add SUSE family for wildcards

Commit:
Fix lint for regexp strings

Commit:
Lintfixes

Commit:
Remove redundant import error capture

Commit:
Add operator checking

Commit:
Add UT for unsupported operators

Commit:
Drop '+' operator

Commit:
Merge pull request #41387 from rallytime/merge-develop

[develop] Merge forward from nitrogen to develop

cachedout commented:
Switch to your copy of the develop branch, pull down the latest from the repo, then switch back to this branch and type `git rebase -i develop`. That will rebase the latest develop onto this branch. You'll then have to force push those changes up to this fork.

Commit:
Merge pull request #41332 from bodhi-space/infra4842

INFRA-4842 - boto_elb should return None when test=True and changes are pending

Commit:
Merge pull request #41400 from raazvvann/heghe/nilrt-fix-example

nilrt_ip: fix examples

Commit:
Merge pull request #41339 from aneeshusa/add-tls-connection-options-for-pgjsonb-returner

Add TLS connection options for pgjsonb returner

Commit:
Fix haproxyconn unit test failures

Too many arguments were being passed into the test function

This change removes the extra "mock" arg and the tests pass.

This issue was introduced in PR #40713

Fixes https://github.com/saltstack/salt-jenkins/issues/357

Commit:
Merge pull request #41405 from isbm/isbm-zypper-wildcards-dev

Wildcards for Zypper

Commit:
Merge pull request #41399 from shitler25oki/develop

Prevent creating cron for user when cron.absent or cron.rm_job is called

Commit:
Merge pull request #41386 from grahamhayes/unit-tests-40905

Add unit test for PR #40905

Commit:
Merge pull request #41294 from hrumph/powershell_runas

Addition of powershell_all function

Commit:
Merge pull request #41411 from rallytime/fix-haproxyconn-tests

Fix haproxyconn unit test failures

Commit:
Merge branch 'salt-states-svn-patch-41022' of https://github.com/sebw/salt into salt-states-svn-patch-41022

cachedout commented:
@sebw It doesn't look like that worked correctly. If it's easier, you can just close this PR and re-open a new one.

sebw commented:
Will do, I tried to follow your instructions but I need to read a little more about rebasing :)
